### PR TITLE
Users

### PR DESCRIPTION
**Pull request by AlexRedisson18**
add modal for password reset
restyle error messages
change default routes at empty views

### CODE REVIEW

**Comment by mkalinin** [comment]
>  disable standart devise routing ( /users/sign_in - will not work )

Yes, you better disable it as it's not needed.

Why change your password has not been implemented in the same fashion as other screens? I mean that it's not rendered as a modal.

**Review by mkalinin — APPROVED** [review]
I've checked Rails part excluding RSpec. It looks good to me 👍

File: Gemfile [hunk 1]
```diff
@@ -8,6 +8,8 @@ gem 'rails', '~> 6.0.2', '>= 6.0.2.2'
 gem 'pg'
 gem 'pry-rails'
 gem 'devise'
+gem 'bootstrap', '~> 4.4.1'
+gem 'jquery-rails'
```
Comment (akurashev) [inline Gemfile]: You don't need this gem. All JS is handled by Webpack and installed through `yarn add ...`.

File: app/assets/stylesheets/devise/modal.sass [hunk 2]
```diff
@@ -0,0 +1,42 @@
+@import url('https://fonts.googleapis.com/css2?family=Gothic+A1:wght@300&display=swap');
+
+*
+  font-family: 'Gothic A1', sans-serif
+
+
+a
+  text-decoration:none !important
```
Comment (akurashev) [inline app/assets/stylesheets/devise/modal.sass]:
```suggestion
  text-decoration: none !important
```

File: app/assets/stylesheets/devise/modal.sass [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+
+#close-modal-cross
+  font-size: 17px
+  color: #005164
+
+#modalSubmitButton
```
Comment (akurashev) [inline app/assets/stylesheets/devise/modal.sass]: Inconsistent naming. I see most of the classes here have `kebab-case` naming, so `modalSubmitButton` -> `modal-submit-button`

File: app/assets/stylesheets/devise/modal.sass [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  color: #005164
+
+.modal-backdrop
+  opacity: 0.1 !important
+
+.modal-content
+  width: 420px !important
+
+
+.modalInput
```
Comment (akurashev) [inline app/assets/stylesheets/devise/modal.sass]:
```suggestion
.modal-input
```

File: app/assets/stylesheets/welcome.scss [hunk 5]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the welcome controller here.
```
Comment (akurashev) [inline app/assets/stylesheets/welcome.scss]: If you don't use it/need it, then just remove it.

File: app/controllers/confirmations_controller.rb [hunk 6]
```diff
@@ -0,0 +1,7 @@
+class ConfirmationsController < Devise::ConfirmationsController
+  private
+
+  def after_confirmation_path_for(resource_name, resource)
+    home_path
+  end
+end
```
Comment (akurashev) [inline app/controllers/confirmations_controller.rb]: Make your editor adding one newline at the end of file automatically.
[Why it matters](https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file)

Do you use Sublime? There's a [package](https://github.com/mattst/sublime-single-trailing-new-line) for it.

File: app/controllers/lots_controller.rb [hunk 7]
```diff
@@ -1,4 +1,7 @@
 class LotsController < ApplicationController
+  # before_action :authenticate_user!
+  # before_action :set_lot, only: %i[show edit uprate destroy]
+
```
Comment (akurashev) [inline app/controllers/lots_controller.rb]: Don't commit commented code.
If you gonna use it in future try [git stash](https://www.atlassian.com/git/tutorials/saving-changes/git-stash)

File: app/javascript/packs/application.js [hunk 8]
```diff
@@ -7,11 +7,79 @@ require("@rails/ujs").start()
 require("turbolinks").start()
 require("@rails/activestorage").start()
 require("channels")
+//= require jquery3
+//= require popper
+//= require bootstrap
+//= require jquery_ujs
+//= require_tree .
```
Comment (akurashev) [inline app/javascript/packs/application.js]: This stuff is for Rails [assets pipeline](https://github.com/rails/sprockets#directives)
But you use Webpack for JS bundling, so they just have no effect. Just remove them.

File: app/javascript/packs/application.js [hunk 9]
```diff
@@ -7,11 +7,79 @@ require("@rails/ujs").start()
...
 // Uncomment to copy all static images under ../images to the output folder and reference
 // them with the image_pack_tag helper in views (e.g <%= image_pack_tag 'rails.png' %>)
 // or the `imagePath` JavaScript helper below.
 //
 // const images = require.context('../images', true)
 // const imagePath = (name) => images(name, true)
+
+
+document.addEventListener('DOMContentLoaded', function() {
+  $("form#sign_in_user")
```
Comment (akurashev) [inline app/javascript/packs/application.js]: Inconsistent ID. `sign_in_user` -> `sign-in-user`

File: app/javascript/packs/application.js [hunk 10]
```diff
@@ -7,11 +7,79 @@ require("@rails/ujs").start()
...
+      const [errors, data, xhr] = event.detail
+      const error_messages =  $.map(errors, function(v, k) {
+        return `<div class='alert alert-danger pull-left'>${v}</div>`;
+      }).join("")
+      return $('.modal').find('.modal-errors').html(error_messages);
+    });
+});
+
+document.addEventListener('DOMContentLoaded', function() {
+  $("form#sign_up_user")
```
Comment (akurashev) [inline app/javascript/packs/application.js]: And here

File: app/javascript/packs/application.js [hunk 11]
```diff
@@ -7,11 +7,79 @@ require("@rails/ujs").start()
...
+      });
+    });
+
+    $('#close-modal-cross').on("click", function(){
+     $(".modal-errors").empty();
+    });
+
+    $('.modal').on("click", function(){
+     $(".modal-errors").empty();
+    });
```
Comment (akurashev) [inline app/javascript/packs/application.js]: Something happened to spacing here :)

File: app/javascript/packs/sign_in.js [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+    error_messages =  $.map(errors, function(v, k) {
+
+      return "<div class='alert alert-danger pull-left'>" + k + " " + v + "</div>";
+
+    }).join("")
+     console.log(error_messages);
+     console.log($(this).parents('.modal').find('.modal-footer'))
+    return $(this).parents('.modal').find('.modal-footer').html(error_messages);
+  });
+});
```
Comment (akurashev) [inline app/javascript/packs/sign_in.js]: Is this used at all? I see almost the same code in `application.js`

File: app/views/layouts/application.html.erb [hunk 13]
```diff
@@ -1,15 +1,26 @@
 <!DOCTYPE html>
 <html>
-  <head>
+    <head>
```
Comment (akurashev) [inline app/views/layouts/application.html.erb]: Why spaces here?

File: app/views/shared/_navbar.html.erb [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+    <% else %>
+      <%= link_to "LOGIN", "", class: 'login-btn', data: { toggle: "modal", target: "#sign_in_modal" } %>
+    <% end %>
+  </form>
+</nav>
+
+<%= render 'shared/devise_modals/sign_in' %>
+<%= render 'shared/devise_modals/sign_up' %>
+<%= render 'shared/devise_modals/new_password' %>
+<%= render 'shared/devise_modals/new_confirmation' %>
```
Comment (akurashev) [inline app/views/shared/_navbar.html.erb]: Modal have nothing to do in navbar. It's better place them in `layouts/application.html.erb` just before the `</body>`

File: app/views/shared/devise_modals/_sign_in.html.erb [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+    </div>
+  </div>
+</div>
+
+<!-- another two form version below -->
+<%# <%= form_for(:user, :url => registration_path(:user), html: { class: 'mx-5 px-4' } ) do |f| %>
+
+<%# <%= form_for(resource, as: resource_name, url: session_path(resource_name)) do |f| %>
+
+<%#<%= form_for(User.new, url: session_path(:user), html:{ id: 'sign_in_user', :'data-type' => 'json', class: 'mx-5 px-4' }, remote: true) do |f| %>
```
Comment (akurashev) [inline app/views/shared/devise_modals/_sign_in.html.erb]: Remove it

File: config/routes.rb [hunk 16]
```diff
@@ -1,5 +1,9 @@
 Rails.application.routes.draw do
-  root 'lots#index'
+  get 'welcome/home'
+
+devise_for :users, controllers: { sessions: 'sessions', registrations: 'registrations', confirmations: 'confirmations' }
```
Comment (akurashev) [inline config/routes.rb]:
```suggestion
  devise_for :users, controllers: { sessions: 'sessions', registrations: 'registrations', confirmations: 'confirmations' }
```

File: index.html [hunk 17]
```diff
@@ -0,0 +1,185 @@
+<!doctype html>
```
Comment (akurashev) [inline index.html]: WTF is this file for?

**Review by akurashev — CHANGES_REQUESTED** [review]

**Comment by akurashev** [comment]
+1 for
> disable standart devise routing ( /users/sign_in - will not work )`

Also remove standard Devise views. They are not used anyway.

And +1 to this question:
> Why change your password has not been implemented in the same fashion as other screens? I mean that it's not rendered as a modal.

File: Gemfile (hunk 1 above)
Comment (AlexRedisson18) [reply Gemfile:12]: dont understand this commit
we have not-separater front and back

File: index.html (hunk 17 above)
Comment (AlexRedisson18) [reply index.html]: it was just bootstrap-code-helper file=D
already deleted

File: app/helpers/welcome_helper.rb [hunk 18]
```diff
@@ -0,0 +1,2 @@
+module WelcomeHelper
```
Comment (onenaviio) [inline app/helpers/welcome_helper.rb]: You don't use this file. Please remove it

File: app/javascript/custom/modal.js [hunk 19]
```diff
@@ -0,0 +1,54 @@
+document.addEventListener('DOMContentLoaded', function() {
+  $("form#sign-in-user")
+  .bind("ajax:success", function(event) {
+    $(this).parents('.modal').modal('hide');
+    location.reload();
+  })
+  .bind("ajax:error", function(event) {
+    const [errors, data, xhr] = event.detail
```
Comment (onenaviio) [inline app/javascript/custom/modal.js]: Second item of `event.detail` array is `status`. Just for know

File: app/javascript/custom/modal.js [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+  });
+
+  $("form#sign-up-user")
+    .bind("ajax:success", function(event) {
+      $(this).parents('.modal').modal('hide');
+      location.reload();
+    })
+    .bind("ajax:error", function(event) {
+      const errors = event.detail[0]["errors"]
+      const errorList = $.map(errors, function(v, k) {
```
Comment (onenaviio) [inline app/javascript/custom/modal.js]: In future try to give more clear names for cycle variables

File: spec/requests/registrations_request_spec.rb [hunk 21]
```diff
@@ -0,0 +1,4 @@
+require 'rails_helper'
+
+RSpec.describe 'Registrations', type: :request do
+end
```
Comment (onenaviio) [inline spec/requests/registrations_request_spec.rb]: I think you don't need this file

File: spec/requests/sessions_request_spec.rb [hunk 22]
```diff
@@ -0,0 +1,4 @@
+require 'rails_helper'
+
+RSpec.describe 'Sessions', type: :request do
+end
```
Comment (onenaviio) [inline spec/requests/sessions_request_spec.rb]: The same

File: spec/requests/welcome_request_spec.rb [hunk 23]
```diff
@@ -0,0 +1,10 @@
+require 'rails_helper'
+
+RSpec.describe 'Welcomes', type: :request do
+  describe 'GET /home' do
+    it 'returns http success' do
+      get '/welcome/home'
+      expect(response).to have_http_status(:success)
+    end
+  end
+end
```
Comment (onenaviio) [inline spec/requests/welcome_request_spec.rb]: try this 
```ruby
RSpec.describe WelcomeController, type: :controller do
  describe 'GET #home' do
    it 'returns http success' do
      get :home
      expect(response).to have_http_status(:success)
    end
  end
end
```

**Review by onenaviio — CHANGES_REQUESTED** [review]

File: app/javascript/custom/modal.js [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+  .bind("ajax:success", function(event) {
+    $(this).parents('.modal').modal('hide');
+    location.reload();
+  })
+  .bind("ajax:error", function(event) {
+    const [errors, status, xhr] = event.detail
+    const error_messages =  $.map(errors, function(value, key) {
+      return `<p class="text-danger">${value}</p>`;
+    }).join("");
+    return $('.modal').find('.modal-errors').html(error_messages);
```
Comment (akurashev) [inline app/javascript/custom/modal.js]:
```suggestion
    return $('#sign-in-user').find('.modal-errors').html(error_messages);
```

File: app/javascript/custom/modal.js [hunk 25]
```diff
@@ -0,0 +1,86 @@
...
+    const checkToken = () => {
+      $(function () {
+        if (searchParams.has('reset_password_token')) {
+          $("#edit-password-modal").modal("show");
+        };
+      const token = searchParams.get('reset_password_token');
+      $("#reset-password-token").val(token);
+      });
+    };
+    checkToken();
```
Comment (akurashev) [inline app/javascript/custom/modal.js]:
```suggestion
    if (searchParams.has('reset_password_token')) {
      $("#edit-password-modal").modal("show");
    }
    const token = searchParams.get('reset_password_token');
    $("#reset-password-token").val(token);
```
Will it work like this?

File: app/javascript/packs/application.js [hunk 26]
```diff
@@ -7,7 +7,9 @@ require("@rails/ujs").start()
 require("turbolinks").start()
 require("@rails/activestorage").start()
 require("channels")
-
+require('jquery')
```
Comment (akurashev) [inline app/javascript/packs/application.js]:
```suggestion
require("jquery")
```
If you started using double quotes then be consistent

**Review by akurashev — CHANGES_REQUESTED** [review]

**Review by akurashev — APPROVED** [review]

**Review by onenaviio — APPROVED** [review]

**Pull request merged**